plugins/inputs/ctrlx_datalayer/ctrlx_datalayer.go

```diff
@@ -0,0 +1,376 @@
+		if err != nil {
+			return nil, err
+		}
+
+		m := metric.New(measurement, tags, flattener.Fields, t)
+		return m, nil
+	case "arbool8",
+		"arint8", "aruint8",
+		"arint16", "aruint16",
+		"arint32", "aruint32",
+		"arint64", "aruint64",
+		"arfloat", "ardouble",
+		"arstring",
+		"artimestamp":
+		fields := make(map[string]interface{})
+		values := em.Value.([]interface{})
+		for i := 0; i < len(values); i++ {
+			index := strconv.Itoa(i)
+			key := fieldKey + "_" + index
+			fields[key] = values[i]
+		}
+		m := metric.New(measurement, tags, fields, t)
+		return m, nil
+	case "bool8",
+		"int8", "uint8",
+		"int16", "uint16",
+		"int32", "uint32",
+		"int64", "uint64",
+		"float", "double",
+		"string",
+		"timestamp":
+		fields := map[string]interface{}{fieldKey: em.Value}
+		m := metric.New(measurement, tags, fields, t)
+		return m, nil
+	}
+
+	return nil, fmt.Errorf("unsupported value type: %s", em.Type)
+}
+
+// Init is for setup, and validating config
+func (c *CtrlXDataLayer) Init() error {
+	// Check all configured subscriptions for valid settings
+	for i := range c.Subscription {
+		sub := &c.Subscription[i]
+		sub.applyDefaultSettings()
+		err := choice.Check(sub.QueueBehaviour, queueBehaviours)
+		if err != nil {
+			c.Log.Infof("The right queue behaviour values are %v", queueBehaviours)
+			return fmt.Errorf("subscription %d: setting 'queue_behaviour' %q is invalid", i, sub.QueueBehaviour)
+		}
+		err = choice.Check(sub.ValueChange, valueChanges)
+		if err != nil {
+			c.Log.Infof("The right value change values are %v", valueChanges)
+			return fmt.Errorf("subscription %d: setting 'value_change' %q is invalid", i, sub.ValueChange)
+		}
+		if len(sub.Nodes) == 0 {
+			c.Log.Warn("A configured subscription has no nodes configured")
+		}
+		sub.index = i
+	}
+
+	// Generate valid communication url based on configured server address
+	u := url.URL{
+		Scheme: "https",
+		Host:   c.Server,
+	}
+	c.url = u.String()
+	_, err := url.Parse(c.url)
+	if err != nil {
+		return errors.New("invalid server address")
+	}
+
+	return nil
+}
+
+// Start input as service, retain the accumulator, establish the connection.
+func (c *CtrlXDataLayer) Start(acc telegraf.Accumulator) error {
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+
+	var err error
+	c.connection, err = c.HTTPClientConfig.CreateClient(ctx, c.Log)
+	if err != nil {
+		return fmt.Errorf("failed to create http client, err: %w", err)
+	}
+
+	username, err := c.Username.Get()
+	if err != nil {
+		return fmt.Errorf("getting username failed: %w", err)
+	}
+	defer config.ReleaseSecret(username)
+
+	password, err := c.Password.Get()
+	if err != nil {
+		return fmt.Errorf("getting password failed: %w", err)
+	}
+	defer config.ReleaseSecret(password)
+
+	c.tokenManager = token.TokenManager{
+		Url:        c.url,
+		Username:   string(username),
+		Password:   string(password),
+		Connection: c.connection,
+	}
```

Comment:
Release the secret as soon as possible, now it would only be done after `gatherLoop` exits.
```suggestion
	username, err := c.Username.Get()
	if err != nil {
		return fmt.Errorf("getting username failed: %w", err)
	}

	password, err := c.Password.Get()
	if err != nil {
		return fmt.Errorf("getting password failed: %w", err)
	}

	c.tokenManager = token.TokenManager{
		Url:        c.url,
		Username:   string(username),
		Password:   string(password),
		Connection: c.connection,
	}
	config.ReleaseSecret(username)
	config.ReleaseSecret(password)
```